liquid-staking/src/liquidity/remove_liquidity.rs

```diff
@@ -0,0 +1,135 @@
+        require!(
+            egld_to_unstake >= MIN_EGLD_TO_DELEGATE,
+            ERROR_INSUFFICIENT_UNSTAKE_AMOUNT
+        );
+        self.burn_ls_token(&payment.amount);
+
+        let delegation_contract = self.get_delegation_contract_for_undelegate(&egld_to_unstake);
+
+        let delegation_contract_mapper = self.delegation_contract_data(&delegation_contract);
+        delegation_contract_mapper
+            .update(|contract_data| contract_data.egld_in_ongoing_undelegation += &egld_to_unstake);
+
+        let gas_for_async_call = self.get_gas_for_async_call();
+        self.tx()
+            .to(delegation_contract.clone())
+            .typed(delegation_proxy::DelegationMockProxy)
+            .undelegate(egld_to_unstake.clone())
+            .gas(gas_for_async_call)
+            .callback(
+                RemoveLiquidityModule::callbacks(self).remove_liquidity_callback(
+                    caller,
+                    delegation_contract,
+                    egld_to_unstake,
+                    payment.amount,
+                ),
+            )
+            .gas_for_callback(MIN_GAS_FOR_CALLBACK)
+            .register_promise();
+    }
+
+    #[promises_callback]
+    fn remove_liquidity_callback(
+        &self,
+        caller: ManagedAddress,
+        delegation_contract: ManagedAddress,
+        egld_to_unstake: BigUint,
+        ls_tokens_to_be_burned: BigUint,
+        #[call_result] result: ManagedAsyncCallResult<()>,
+    ) {
+        let mut storage_cache = StorageCache::new(self);
+        let delegation_contract_mapper = self.delegation_contract_data(&delegation_contract);
+
+        delegation_contract_mapper.update(|contract_data| {
+            contract_data.egld_in_ongoing_undelegation -= &egld_to_unstake;
+        });
+
+        match result {
+            ManagedAsyncCallResult::Ok(()) => {
+                let current_epoch = self.blockchain().get_block_epoch();
+                let unbond_epoch = current_epoch + UNBOND_PERIOD;
+
+                delegation_contract_mapper.update(|contract_data| {
+                    contract_data.total_staked_from_ls_contract -= &egld_to_unstake;
+                    contract_data.total_unstaked_from_ls_contract += &egld_to_unstake;
+                });
+
+                let virtual_position = UnstakeTokenAttributes {
+                    delegation_contract,
+                    unstake_epoch: current_epoch,
+                    unstake_amount: egld_to_unstake,
+                    unbond_epoch,
+                };
+
+                let user_payment = self.mint_unstake_tokens(&virtual_position);
+                self.send().direct_esdt(
+                    &caller,
+                    &user_payment.token_identifier,
+                    user_payment.token_nonce,
+                    &user_payment.amount,
+                );
+
+                self.emit_remove_liquidity_event(
```

Comment:
You only appear to use `storage_cache` here, so move it's definition right before this line.